connectors/memory/memory_test.go

```diff
@@ -0,0 +1,650 @@
+			{Name: "f1", Type: dosa.String},
+			{Name: "c1", Type: dosa.Int64},
+			{Name: "c2", Type: dosa.Double},
+			{Name: "c3", Type: dosa.String},
+			{Name: "c4", Type: dosa.Blob},
+			{Name: "c5", Type: dosa.Bool},
+			{Name: "c6", Type: dosa.Int32},
+			{Name: "c7", Type: dosa.TUUID},
+		},
+		Key: &dosa.PrimaryKey{
+			PartitionKeys: []string{"f1"},
+		},
+		Name: "t1",
+	},
+}
+var clusteredEi = &dosa.EntityInfo{
+	Ref: &testSchemaRef,
+	Def: &dosa.EntityDefinition{
+		Columns: []*dosa.ColumnDefinition{
+			{Name: "f1", Type: dosa.String},
+			{Name: "c1", Type: dosa.Int64},
+			{Name: "c2", Type: dosa.Double},
+			{Name: "c3", Type: dosa.String},
+			{Name: "c4", Type: dosa.Blob},
+			{Name: "c5", Type: dosa.Bool},
+			{Name: "c6", Type: dosa.Int32},
+			{Name: "c7", Type: dosa.TUUID},
+		},
+		Key: &dosa.PrimaryKey{
+			PartitionKeys: []string{"f1"},
+			ClusteringKeys: []*dosa.ClusteringKey{
+				{Name: "c1", Descending: false},
+				{Name: "c7", Descending: true},
+			},
+		},
+		Name: "t1",
+	},
+}
+
+func TestConnector_CreateIfNotExists(t *testing.T) {
+	sut := Connector{}
+
+	err := sut.CreateIfNotExists(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data"),
+	})
+	assert.NoError(t, err)
+
+	err = sut.CreateIfNotExists(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data"),
+	})
+
+	assert.Error(t, err)
+	assert.True(t, dosa.ErrorIsAlreadyExists(err))
+}
+func TestConnector_Upsert(t *testing.T) {
+	sut := Connector{}
+
+	err := sut.Upsert(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data"),
+	})
+	assert.NoError(t, err)
+	vals, err := sut.Read(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data")}, []string{"c1"})
+	assert.NoError(t, err)
+	assert.Nil(t, vals["c1"])
+
+	err = sut.Upsert(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data"),
+		"c1": dosa.FieldValue(int64(1)),
+	})
+	assert.NoError(t, err)
+
+	vals, err = sut.Read(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data")}, []string{"c1"})
+	assert.NoError(t, err)
+	assert.Equal(t, dosa.FieldValue(int64(1)), vals["c1"])
+}
+
+func TestConnector_Read(t *testing.T) {
+	sut := Connector{}
+
+	// read with no data
+	vals, err := sut.Read(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data")}, []string{"c1"})
+	assert.Error(t, err)
+	assert.True(t, dosa.ErrorIsNotFound(err))
+
+	err = sut.CreateIfNotExists(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data"),
+		"c1": dosa.FieldValue(int64(1)),
+		"c2": dosa.FieldValue(float64(2)),
+	})
+	assert.NoError(t, err)
+
+	vals, err = sut.Read(context.TODO(), testEi, map[string]dosa.FieldValue{
+		"f1": dosa.FieldValue("data")}, []string{"c1"})
+	assert.NoError(t, err)
+	assert.Equal(t, int64(1), vals["c1"])
+	assert.Nil(t, vals["c2"])
+	assert.Equal(t, 1, len(vals))
```

Comment:
nit: the assert package allows you to semantically assert length like so:
```
assert.Len(t, vals, 1)
```

Save you a few chars